rest/rest-server/src/main/java/org/ow2/proactive_grid_cloud_portal/scheduler/SchedulerStateRest.java

```diff
@@ -2692,16 +2706,27 @@ public Response index() throws URISyntaxException {
     }
 
     private String downloadWorkflowContent(String sessionId, String workflowUrl)
-            throws JobCreationRestException, IOException {
-        if (StringUtils.isBlank(workflowUrl))
+            throws JobCreationRestException, IOException, NotConnectedRestException {
+        if (StringUtils.isBlank(workflowUrl)) {
             throw new JobCreationRestException("Cannot create workflow without url");
+        }
+        // cache key contains the owner and the workflow url, as different owners may have different read access rights
+        String userName = getUserName(sessionId);
+        String key = userName + "_" + workflowUrl;
+        String cachedWorkflow = workflowCache.getIfPresent(key);
+        if (cachedWorkflow != null) {
+            return cachedWorkflow;
+        }
+        String downloadedWorkflow = null;
         if (workflowUrl.startsWith("http")) {
             HttpResourceDownloader httpResourceDownloader = HttpResourceDownloader.getInstance();
-            return httpResourceDownloader.getResource(sessionId, workflowUrl, String.class);
+            downloadedWorkflow = httpResourceDownloader.getResource(sessionId, workflowUrl, String.class);
         } else {
             URL nonHttpURL = new URL(workflowUrl);
-            return IOUtils.toString(nonHttpURL.openStream(), Charsets.UTF_8);
+            downloadedWorkflow = IOUtils.toString(nonHttpURL.openStream(), Charsets.UTF_8);
         }
+        workflowCache.put(key, downloadedWorkflow);
+        return downloadedWorkflow;
```

Comment:
Should we check that 'downloadedWorkflow' is not null before returning it ? otherwise throw an exception

Reply:
httpResourceDownloader.getResource() and downloadedWorkflow = IOUtils.toString() don't seem to be able to return null, they would throw exceptions. But I can add a null check and throw a meaningful exception just in case